xarray/coding/times.py

```diff
@@ -203,18 +238,50 @@ def _infer_time_units_from_diff(unique_timedeltas):
     return 'seconds'
 
 
+def infer_calendar_name(dates):
+    """Given an array of datetimes, infer the CF calendar name"""
+    if np.asarray(dates).dtype == 'datetime64[ns]':
+        return 'proleptic_gregorian'
+    else:
+        try:
+            return np.asarray(dates)[0].calendar
+        except IndexError:
+            return np.asarray(dates).item().calendar
+
+
 def infer_datetime_units(dates):
     """Given an array of datetimes, returns a CF compatible time-unit string of
     the form "{time_unit} since {date[0]}", where `time_unit` is 'days',
     'hours', 'minutes' or 'seconds' (the first one that can evenly divide all
     unique time deltas in `dates`)
     """
-    dates = pd.to_datetime(np.asarray(dates).ravel(), box=False)
-    dates = dates[pd.notnull(dates)]
-    unique_timedeltas = np.unique(np.diff(dates))
+    if np.asarray(dates).dtype == 'datetime64[ns]':
+        dates = pd.to_datetime(np.asarray(dates).ravel(), box=False)
+        dates = dates[pd.notnull(dates)]
+        unique_timedeltas = np.unique(np.diff(dates))
+        reference_date = dates[0] if len(dates) > 0 else '1970-01-01'
+        reference_date = pd.Timestamp(reference_date)
+    else:
+        if not OPTIONS['enable_cftimeindex']:
+            raise ValueError('Serializing dates of type cftime.datetime '
+                             'requires setting enable_cftimeindex to True and '
+                             'using the standalone cftime library to enable '
+                             'accurate roundtripping of date types.')
+        dates = np.asarray(dates).ravel()
+        unique_timedeltas = np.unique(pd.to_timedelta(np.diff(dates)))
```

Comment:
I'm surprised --  `pd.to_timedelta` works to convert cftime.datetime objects?

Reply:
Calling `np.diff` on the dates converts things to generic `datetime.timedelta` objects before they are passed to `pd.to_timedelta`.

Reply:
OK, that makes sense. It still doesn't quite make sense to me why we need `pd.to_timedelta()` here but not with the non-cftime path.

Reply:
We need to convert from `datetime.timedelta` objects to `np.timedelta64` objects to prepare things for `_infer_time_units_from_diff`, which requires `np.timedelta64` objects.  Upon looking at this again, it seems the following also works for this type conversion:
```python
unique_timedeltas = np.unique(np.diff(dates)).astype('timedelta64[ns]')
```
Would you prefer that solution?

Reply:
Yes, I think that's clearer.

Reply:
Great, I went with that in 3947aac.